### ISSUE TITLE
add validation routines / strategies to facilitate generating valid EML?

### ISSUE THREAD

**Issue by cboettig** [ropensci/EML#147]
As suggested by @mbjones in https://github.com/ropensci/EML/issues/140#issuecomment-192517186

Writing S4 validate methods may be too strict; in particular, it may sometimes be easier to create a new element that doesn't yet have all the required metadata, and then add that metadata later.  A validation method would object immediately, making this strategy harder.

Still, it would be super nice to have a better way to discover that a given element is missing some slots that are required.  We might do best to write custom `validate()` methods for at least the more commonly created slots to help with this. 

`eml_validate()` can help, and notably it can be called on a sub-element if the appropriate namespace is given:

```r
dataset <- new("dataset", title = "incomplete, invalid EML")
v <- eml_validate(dataset, namespaces = c(ds = "eml://ecoinformatics.org/dataset-2.1.1"), ns = "ds")
v$errors[[1]]$msg 
```

shows:

```
[1] "Element '{eml://ecoinformatics.org/dataset-2.1.1}dataset': Missing child element(s). Expected is one of ( title, creator ).\n"
```

Shows that this is missing a `creator` element (though admittedly one might think the error is saying it is missing a `title` element too, which it has... XMLSchemaValidate / libxml2 validator may have some limitations....)

This is nice in that it validates against the schema directly instead of some manually-translated set of rules from the schema, but is not very convenient.  I do find the error messages coming out of `XMLSchemaValidate` function often more readable than those that come from the online eml parser (benefits from real line numbers and not tons of `(((((((`.   Yes, I know the latter is more complete in checking reference/ids too), but `XMLSchemaValidate` has the down-side of stopping at the first error message.

**Comment by ivanhanigan** [comment]
Query: does this include looking at the data too.  So if the
geographicCoverage specifies a particular projection, or an extent, the
validation would involve checking that the data are actually projected in
that coordinate system and have that extent?  I have lost count of the
number of NetCDF climate grids that either are projected but don't declare
it in the metadata, or vice verca.  Also applies to temporalCoverage where
Date data could be checked to see if it really does fall in the range
specified by the EML, or for numeric data that the values do conform to
specified maxima and minima?

I ask because I suspect you could get a few valid EML documents that have
inconsistencies with the data entities they refer to.

**Comment by cboettig** [comment]
@ivanhanigan Good question.  Once we have some routines that help automatically generate, say, geographicCoverage data from actual data files (as discussed in #150), it would be relatively straight forward to make such a comparison.  Might be better to discuss in that thread.  

In this issue, I'm specifically concerned only with generating EML that meets the validation criteria of the schema, which is somewhat of a separate issue.  Ensuring that data is provided in a manner consistent with the EML schema makes it possible to write software tools & scripts that can reliably parse the metadata to begin with. I would suggest that any applications using EML data must still contend with issues such as the inaccuracies you mention, and may employ a range of other data-validation techniques as well (e.g. that locations that are supposed to be terrestrial aren't in the ocean, etc), just as with working with any other data.  But we have to walk before we can run: the scope of EML is to make it possible in the first place to design such tools.  If the software doesn't know how to find the right geographic coordinates in the first place (which is more likely when parsing some format without a clearly defined schema), than it's harder to build such validation tools.  

So I think the things you suggest are great use cases for applications that use EML, but may be beyond the scope of the EML package right now.

**Comment by mbjones** [comment]
@ivanhanigan I agree with @cboettig that data parsing and validation is beyond the scope of the EML package, but I do have another project with Lindsay Powers and Ted Habermann to work on just that ability.  The @LTER network already has an EML quality checker that is part of their [PASTA](https://github.com/lter/PASTA) framework.  We will be extending this quality engine to include more metadata schemas, an easily configurable set of rules, and a nice web-based reporting framework to display the results in the [KNB](https://knb.ecoinformatics.org/#data) and [DataONE](https://search.dataone.org) portals.  Rules that are checked are currently written in Java, but I am considering adding in an R engine to the service to allow custom quality check rules to be written in R and maybe python.  LTER has a [quality report in XML format](https://cn.dataone.org/cn/v2/resolve/https%3A%2F%2Fpasta.lternet.edu%2Fpackage%2Freport%2Feml%2Fknb-lter-hbr%2F76%2F6) for many or most (or all) of their data sets in PASTA if you want to see some example reports.  This is the starting point for our new project which will generalize and broaden this framework, and make it available across [DataONE](https://search.dataone.org).

**Comment by cboettig** [comment]
PR #154 gives us a bit prettier validation results.  Class definitions and `get` / `set` functions are designed to promote valid EML, though it may always be possible to create invalid EML if you really want to do so.  Meanwhile think we can close this and open actionable issues for specific examples where R functions most risk generating invalid EML.
